i18n-helpers/src/bin/markdown-gettext.rs

```diff
@@ -0,0 +1,250 @@
+use mdbook_i18n_helpers::translate;
+use polib::catalog::Catalog;
+use polib::po_file;
+use std::fs;
+use std::fs::File;
+use std::io::{BufReader, Read, Write};
+use std::path::{Path, PathBuf};
+
+fn build_catalog(lang: &str) -> anyhow::Result<Catalog> {
+    let pot_path = Path::new(lang);
+    let catalog = po_file::parse(pot_path)
+        .map_err(|err| anyhow!("{err}"))
+        .with_context(|| format!("Could not parse {:?} as PO file", pot_path))?;
+
+    Ok(catalog)
+}
+
+fn translate_files(
+    catalog: Catalog,
+    files: Vec<PathBuf>,
+    output_path: PathBuf,
+) -> anyhow::Result<()> {
+    for file_path in files.iter() {
+        let content = File::open(file_path)
+            .with_context(|| format!("Failed to open file: {}", file_path.display()))?;
+        let mut buf_reader = BufReader::new(content);
+        let mut contents = String::new();
+        buf_reader
+            .read_to_string(&mut contents)
+            .with_context(|| format!("Failed to read file: {}", file_path.display()))?;
+
+        let translated_content = translate(&contents, &catalog);
+
+        let output_file_name = file_path.file_name().ok_or(anyhow!("Invalid file name"))?;
+        let output_file_path = output_path.join(output_file_name);
+
+        let mut output_file = File::create(&output_file_path)
+            .with_context(|| format!("Failed to create file: {}", output_file_path.display()))?;
+
+        output_file
+            .write_all(translated_content.as_bytes())
+            .with_context(|| format!("Failed to write to file: {}", output_file_path.display()))?;
+    }
+    Ok(())
+}
+
+fn validate_file(path: &Path) -> bool {
+    path.is_file() && path.extension().map_or(false, |ext| ext == "md")
+}
+
+fn allocate_files(
+    dir_option: Option<&str>,
+    file_option: Option<&str>,
+) -> anyhow::Result<Vec<PathBuf>> {
+    let mut valid_files = Vec::new();
+    match (dir_option, file_option) {
+        (Some(dir_path), None) => {
+            let full_dir = PathBuf::from(dir_path);
+            fs::read_dir(full_dir)?
+                .filter_map(Result::ok)
+                .filter(|entry| validate_file(&entry.path()))
+                .for_each(|entry| valid_files.push(entry.path()));
+        }
+        (None, Some(file_path)) => {
+            let full_file_path = PathBuf::from(file_path);
+            if !validate_file(&full_file_path) {
+                bail!("Markdown file does not exist: {}", full_file_path.display())
+            }
+            valid_files.push(full_file_path);
+        }
+        _ => unreachable!(),
+    }
+    Ok(valid_files)
+}
+
+fn build_output_path(output_path_option: Option<&str>) -> anyhow::Result<PathBuf> {
+    let output_path = output_path_option.unwrap_or("translated_md_files");
+    let path = PathBuf::from(output_path);
+    fs::create_dir_all(&path)
+        .with_context(|| format!("Failed to create directory: {}", path.display()))?;
+    Ok(path)
+}
+
+fn main() -> anyhow::Result<()> {
```

Comment:
I'm a bit in doubt about the best command line interface for this...

One idea would be to allow

```shell
markdown-gettext xx.po foo.md -o foo-xx.md
```

which would translate `foo.md` using `xx.po` and write the output to `foo-xx.md`.

Here, `foo.md` could be left out or it could be `-` and then we read from `stdin` (like for `markdown-xgettext` below). Similar, `-o` can be left out or be `-` and then we write to `stdout`.

If there are multiple inputs, then I guess the user cannot write to `stdout` and must instead give an output directory:

```shell
markdown-gettext xx.po foo.md bar.md --output-dir xx
```

In this case it would be an error to attempt to read from `stdin` since we don't really know what to name the output file in that case.

Reply:
If I'm understanding correctly, we want optional flags for output file title and directory. 

Since `markdown-gettext xx.po foo.md -o foo-xx.md`, does this mean if one was to add an additional wef.md, would the command be: `markdown-gettext xx.po foo.md -o foo-xx.md wef-xx.md`?

Output directory makes sense to me 👍. I would just like some clarification on file naming conventions, especially when one wants to translate multiple files.

Reply:
> If I'm understanding correctly, we want optional flags for output file title and directory.

Yeah, I'm trying to figure out how we can model the behavior here to be similar to existing Unix commands.

We should also keep it simple! Less code for all of us to write and maintain :smile: 

I'm thinking of using [`cp`](https://man7.org/linux/man-pages/man1/cp.1.html) as a basis:

- `cp` can copy a single path to a new name _or_
- `cp` can copy multiple path into a single directory

The paths which are copied can be files or directories (or a mix of both). If there is a directory, then you must add `-r` to do a recursive copy.

I now see

```
markdown-gettext xx.po source.md dest.md
```
as

```
cp source.md dest.md
```
but with the translation happening as well.

Similarly, 

```
markdown-gettext xx.po foo.md bar.md dest/
```
as

```
cp foo.md bar.md dest/
```

Finally, the rule out `-r` should be implemented: if you want to recursively translate a directory, you have to add a `-r` flag.
 
This new thinking is different from what I had before where I thought we should use the flags from the GNU Gettext utilities. Back then, I hadn't realized that there isn't a command line program which mirrors what our `markdown-gettext` program will do :smile:

Note that these are just ideas, please let me know if you have better ones!

Reply:
@mgeisler I'll write up a doc that outlines what the CLI should do for both `xgettext` and `gettext`. Once we get everything to your liking, I'll change the binaries to follow those requirements!

Reply:
Sounds great! That document could simultaneously become the manual for the tools, so your work here will be reused right away.

Reply:
Hey @friendlymatthew, sorry for the slow reply. The above looks great, thank you very much for writing it up! I think it follow the behavior from the [`cp` manual page](https://man7.org/linux/man-pages/man1/cp.1.html) closely.

Reply:
No worries @mgeisler, I appreciate the opportunity to work on this 😁. 

When I have some time off from official work, I'll get on this ASAP